### ISSUE TITLE
BUILTIN NATURAL internal error at Forcing.hs:232

### ISSUE THREAD

**Issue by laMudri** [agda/agda#3930]
Checking this example in Agda 2.6.0.1
```agda
module issue3930 where

  data ℕ : Set where
    zero : ℕ
    suc : ℕ → ℕ
  {-# BUILTIN NATURAL ℕ #-}

  _+_ : (m n : ℕ) → ℕ
  zero + n = n
  suc m + n = suc (m + n)

  data Th : (m n : ℕ) → Set where
    os : ∀ {m n} → Th m n → Th (suc m) (suc n)

  Fin : ℕ → Set
  Fin = Th (suc zero)

  infixl 6 _+↓_
  infix 4 _≈M_

  postulate
    U : Set
    RCtx : ℕ → Set
    _+↓_ : ∀ {m n} → RCtx m → RCtx n → RCtx (n + m)
    [-_-] : U → RCtx (suc zero)
    _≈M_ : ∀ {n} (Δ0 Δ1 : RCtx n) → Set

  infix 4 _⊢l-var_

  data _⊢l-var_ : ∀ {n} (Δi : RCtx n) (i : Fin n) → Set where
    os : ∀ {n} {e : Th zero n} {Δ Δπ π} (iq : Δπ ≈M Δ +↓ [- π -]) →
         Δπ ⊢l-var os e

  ⊢l-var-sub : ∀ {n Δ} {i : Fin n} → Δ ⊢l-var i → Set
  ⊢l-var-sub (os iq) = {!!}
```
gives error message
```
An internal error has occurred. Please report this as a bug.
Location of the error: src/full/Agda/TypeChecking/Forcing.hs:232
```

Note that the error only turns up at the pattern match, not at the data declaration.

Some experiments:

* Removing the pragma `  {-# BUILTIN NATURAL ℕ #-}` makes it check fine.
* Replacing all of the expressions like `Δ +↓ [- π -]` using a new postulate `_-,_ : RCtx m → U → RCtx (suc m)`, resulting in `Δ -, π` makes it check fine.
* Using `{-# BUILTIN NATPLUS _+_ #-}` makes no difference.
* Removing the index `i : Fin n` makes everything check fine.

**Comment by gallais** [comment]
Thanks for the shrunk down test case!

Removed the `variable` block so that it could be tested with older versions of
Agda. Works fine on 2.5.4.2.

**Comment by andreasabel** [comment]
This could be related to #3898.

**Comment by UlfNorell** [comment]
Problem is indeed with reduction, but not really related to #3898. The problem there is during the analysis of which arguments are forced. In this case the problem is the translation of left-hand sides to rebind patterns in forced positions. Fix: #4028.
